### PR TITLE
Add rewards notification to backup wallet

### PR DESCRIPTION
**Pull request by emerick**
Fixes brave/brave-browser#1164
Requires brave/brave-ui#312

Notify the user that they need to backup their wallet if they haven't done so for 7 days after funding it. Backup can be completed by going to the wallet properties in Rewards settings and selecting either Copy, Print, or Save to File. If notification is ignored, another notification is shown every 7 days.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [x] Windows
  - [ ] macOS
  - [ ] Linux
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

Each test should start with a clean profile.

- You can set `brave.rewards.backup_notification_frequency` (units are in microseconds) in your profile's `Preferences` file to shorten the backup notification interval (defaults to 7 days).
- You can set `brave.rewards.notification_timer_interval` (units are in microseconds) in your profile's `Preferences` file to shorten the interval that we wait before checking if notifications need to be shown (defaults to 15 minutes).

### Test 1

- Delete any existing profile
- Launch Brave
- Enable Rewards
- Don't add any funds to wallet
- Wait 7 days (if you don't have 7 days to spare, see tip above for shortening this interval)
- Now wait > 15 minutes for notification check (or see tips above to shorten this interval)
- Rewards icon should _not_ be badged
- `brave.rewards.backup_succeeded` should be set to `false` in `Preferences` for profile

### Test 2

- Delete any existing profile
- Launch Brave
- Enable Rewards
- Add funds to wallet
- Wait 7 days (if you don't have 7 days to spare, see tip above for shortening this interval)
- Now wait > 15 minutes for notification check (or see tips above to shorten this interval)
- Rewards icon should be badged and should contain a notification like this:

image

- Clear notification
- Bring up wallet properties dialog and select Copy or Print or Save to File
- Wait 7 days again (if you don't have 7 days to spare, see tip above for shortening this interval)
- Wait > 15 minutes again for next notification check (or see tips above to shorten this interval)
- Rewards icon should _not_ rebadge
- `brave.rewards.backup_succeeded` should be set to `true` in `Preferences` file for profile

### Test 3

- Delete any existing profile
- Launch Brave
- Enable Rewards
- Add funds to wallet
- Wait 7 days (if you don't have 7 days to spare, see tip above for shortening this interval)
- Now wait > 15 minutes for notification check (or see tips above to shorten this interval)
- Rewards icon should be badged and should contain a notification like this:

image

- Clear notification
- Don't backup wallet
- Wait 7 days again (if you don't have 7 days to spare, see tip above for shortening this interval)
- Wait > 15 minutes again for next notification check (or see tips above to shorten this interval)
- Rewards icon _should_ rebadge
- `brave.rewards.backup_succeeded` should be set to `false` in `Preferences` file for profile

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by emerick** [comment]
@NejcZdovc Ah, yes it won't be populated initially until the timer fires _and_ the user has funded. I didn't realize that the default settings wouldn't appear in preferences on startup. I could make the code do that on startup or you can just add them manually for testing purposes, just let me know which you prefer.

File: components/brave_rewards/browser/rewards_service.cc [hunk 1]
```diff
@@ -41,6 +42,15 @@ void RewardsService::RemoveObserver(RewardsServiceObserver* observer) {
 // static
 void RewardsService::RegisterProfilePrefs(PrefRegistrySimple* registry) {
   registry->RegisterStringPref(kRewardsNotifications, "");
+  registry->RegisterTimeDeltaPref(kRewardsNotificationTimerInterval,
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service.cc]: I think this one is not used anymore

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 2]
```diff
@@ -338,6 +339,38 @@ void RewardsServiceImpl::Init() {
   ledger_->Initialize();
 }
 
+void RewardsServiceImpl::MaybeShowBackupNotification() {
+  PrefService* pref_service = profile_->GetPrefs();
+  bool user_has_funded = pref_service->GetBoolean(kRewardsUserHasFunded);
+  bool backup_succeeded = pref_service->GetBoolean(kRewardsBackupSucceeded);
+  if (user_has_funded && !backup_succeeded) {
+    base::Time now = base::Time::Now();
+    base::Time boot_timestamp = pref_service->GetTime(kRewardsBootTimestamp);
```
Comment (NejcZdovc) [inline components/brave_rewards/browser/rewards_service_impl.cc]: we have boot time in native ledger. I think we should use that one. Currently we don't have it exposed, but we can add getter for it

File: components/brave_rewards/browser/rewards_service.cc (hunk 1 above)
Comment (emerick) [reply components/brave_rewards/browser/rewards_service.cc]: Yep, I added it and forgot to use it. Fixed!

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 2 above)
Comment (emerick) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Fixed.

File: components/brave_rewards/resources/extension/brave_rewards/background/api/locale_api.ts [hunk 3]
```diff
@@ -22,6 +22,7 @@ export const getMessage = (message: string, substitutions?: string[]): string =>
 
 export const getUIMessages = (): Record<string, string> => {
   const strings = [
+    'backupWalletTitle',
```
Comment (NejcZdovc) [inline components/brave_rewards/resources/extension/brave_rewards/background/api/locale_api.ts]: you need to add `backupWalletNotification` here as well

Comment (emerick) [reply components/brave_rewards/resources/extension/brave_rewards/background/api/locale_api.ts]: Fixed, and removed translation from `en_GB`.

**Comment by LaurenWags** [comment]
@emerick is Test 1 still valid based on https://github.com/brave/brave-core/pull/1044#issuecomment-447862436?

**Comment by emerick** [comment]
@LaurenWags It's still valid, it's just that you have to manually add the notification preferences to `Local State` as they don't appear there on startup. I can hunt up the settings to use there if it's helpful!

**Pull request merged**